### PR TITLE
Raise the correct exception in fast_serialize_string

### PR DESCRIPTION
**Pull request by eregon**
Fix: https://github.com/ruby/json/issues/634

* Related to https://github.com/ruby/json/issues/344

### CODE REVIEW

File: lib/json/pure/generator.rb [hunk 1]
```diff
@@ -339,6 +339,7 @@ def generate(obj)
           private def fast_serialize_string(string, buf) # :nodoc:
             buf << '"'
             string = string.encode(::Encoding::UTF_8) unless string.encoding == ::Encoding::UTF_8
+            raise GeneratorError, "source sequence is illegal/malformed utf-8" unless string.valid_encoding?
```
Comment (byroot) [inline lib/json/pure/generator.rb:342]: Probably need to do it in the Ruby 2.3 version too?

File: test/json/json_generator_test.rb [hunk 2]
```diff
@@ -443,6 +443,20 @@ def test_invalid_encoding_string
...
+    assert_includes error.message, "source sequence is illegal/malformed utf-8"
+
+    # These pass on the pure-Ruby generator but not with the native extension
+    # assert_raise(Encoding::UndefinedConversionError) do
+    #   "\x82\xAC\xEF".b.to_json
+    # end
+    #
+    # assert_raise(Encoding::UndefinedConversionError) do
+    #   JSON.dump("\x82\xAC\xEF".b)
+    # end
```
Comment (byroot) [inline test/json/json_generator_test.rb:459]: You can conditionally run it with `if defined?(JSON::Pure)`.

I'll be trying to make implementations converge over time, but short term it's helpful to test the divergent behavior.

Comment (eregon) [inline test/json/json_generator_test.rb:459]: I filed https://github.com/ruby/json/issues/634 for this

File: lib/json/pure/generator.rb (hunk 1 above)
Comment (eregon) [reply lib/json/pure/generator.rb:342]: That one calls `String#to_json` so it's fine

File: test/json/json_generator_test.rb (hunk 2 above)
Comment (eregon) [reply test/json/json_generator_test.rb]: OK, done.

Comment (byroot) [reply test/json/json_generator_test.rb]: Actually there is already a test for it, so just remove the commented out code I think: https://github.com/ruby/json/blob/f887daa403a7f5ad880c48ef3f6c72615df08a80/test/json/json_generator_test.rb#L448-L456

Comment (eregon) [reply test/json/json_generator_test.rb]: That's different, it's for valid UTF-8 bytes, this is invalid UTF-8 bytes.

**Pull request merged**